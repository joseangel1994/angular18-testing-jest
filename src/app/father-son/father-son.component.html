<h5>Son</h5>
<code id="client-code">{{ client | json}}</code>
<br>

@if (client) {
  <button id="btn-change" (click)="onChange(5)">Cambiar ID</button>
  <button id="btn-delete" (click)="onDelete()">Borrar cliente</button>
}
