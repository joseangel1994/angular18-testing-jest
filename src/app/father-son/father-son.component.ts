import {Component, EventEmitter, Input, Output} from '@angular/core';
import {JsonPipe} from "@angular/common";
import {Client} from "../interfaces";

@Component({
  selector: 'app-father-son',
  standalone: true,
  imports: [
    JsonPipe
  ],
  templateUrl: './father-son.component.html'
})
export class FatherSonComponent {

  @Input() client?: Client
  @Output() onDeleteClient = new EventEmitter()
  @Output() onClientUpdated = new EventEmitter<Client>()

  onDelete() {
    this.client = undefined
    this.onDeleteClient.emit()
  }

  onChange(newId: number) {
    if(!this.client) return

    this.client = {
      ...this.client,
      id: newId
    }

    this.onClientUpdated.emit(this.client)
  }

}
