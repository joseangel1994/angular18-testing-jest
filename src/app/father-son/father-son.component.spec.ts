import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherSonComponent } from './father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [FatherSonComponent]
    })
    .compileComponents();

    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement
    jest.clearAllMocks()
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(compiled).toMatchSnapshot()
  })

  it('buttons should not appear if no client', () => {
    const buttons = compiled.querySelectorAll('button')
    expect(buttons.length).toBe(0)
  })

  it('2 buttons should appear if client', () => {
    component.client = {id: 1, name: 'Jose'}
    fixture.detectChanges()

    const buttons = compiled.querySelectorAll('button')
    expect(buttons.length).toBe(2)
  })

  it('should emit ondeleteclient after click on button', () => {
    component.client = {id: 1, name: 'Jose'}
    fixture.detectChanges()

    jest.spyOn(component.onDeleteClient, 'emit')

    const btnDelete: HTMLButtonElement | null = compiled.querySelector('#btn-delete')
    btnDelete?.click()

    expect(component.onDeleteClient.emit).toHaveBeenCalled()
  })

  it('should emit onclientupdated after click on button', () => {
    component.client = {id: 1, name: 'Jose'}
    fixture.detectChanges()

    jest.spyOn(component.onClientUpdated, 'emit')

    const btnChange: HTMLButtonElement | null = compiled.querySelector('#btn-change')
    btnChange?.click()

    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({
      id: 5,
      name: 'Jose'
    })
  })

  it('should emit onclientupdated if there is client', () => {
    jest.spyOn(component.onClientUpdated, 'emit')
    component.onChange(10)

    expect(component.onClientUpdated.emit).not.toHaveBeenCalled()

    component.client = {id: 1, name: 'Jose'}
    fixture.detectChanges()
    component.onChange(10)

    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({
      id: 10, name: 'Jose'
    })

  })
});
