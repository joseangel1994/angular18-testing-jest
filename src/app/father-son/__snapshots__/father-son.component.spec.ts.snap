// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Son
  </h5>
  <code>
    
  </code>
  <br />
  
</div>
`;
